### ISSUE TITLE
Dependencies on PyPI needs to be updated. 

### ISSUE THREAD

**Issue by alvations** [nltk/nltk#1531]
Most users installs `nltk` through:

```
$ pip install -U nltk
```

or 

```
$ conda install nltk
```

but somehow the PyPI requirements seems to be far lacking behind the ones in `pip-req.txt`.

**Comment by fievelk** [comment]
I think this is related to the fact that [`nltk/setup.py`][1] only lists `six` as required dependency:
```python
setup(
    ...
    install_requires = ['six>=1.10.0'],
    ...
)
```

From [`pip` docs][2]:
> **`Install_requires`**
> `install_requires` is a setuptools `setup.py` keyword that should be used to specify what a project minimally needs to run correctly. When the project is installed by `pip`, this is the specification that is used to install its dependencies.

Conversely, [`nltk/pip-req.txt`][3] is a [requirement file][4]:
> Whereas `install_requires` metadata is automatically analyzed by `pip` during an install, requirements files are not, and only are used when a user specifically installs them using `pip install -r`.

(See [docs][5]).

Please note that [`nltk/jenkins.sh`][6] is actually using `pip-req.txt`:
```bash
#download nltk python dependencies
pip install --upgrade -r pip-req.txt --allow-external matplotlib --allow-unverified matplotlib
```

So I think we should add our minimal requirements to `install_requires` in `setup.py`. Moreover, in order to avoid redundancy, we could probably think about completely removing `pip-req.txt`.

@stevenbird @kmike what do you think about this?

[1]: https://github.com/nltk/nltk/blob/4a4a5031f5c0f99b1ca13079518c221b338ef413/setup.py
[2]: https://packaging.python.org/requirements/#install-requires
[3]: https://github.com/nltk/nltk/blob/4a4a5031f5c0f99b1ca13079518c221b338ef413/pip-req.txt
[4]: https://pip.pypa.io/en/latest/user_guide/#requirements-files
[5]: https://packaging.python.org/requirements/#requirements-files
[6]: https://github.com/nltk/nltk/blob/4a4a5031f5c0f99b1ca13079518c221b338ef413/jenkins.sh

**Comment by kmike** [comment]
I think if NLTK is usable without matplotlib then we shouldn't add matplotlib to install_requires. It is a huge dependency, and there is no reason to force users to install it. Creating extra_requires makes more sense - this way user may e.g. run `pip install nltk[plotting]` and get all dependencies required for nltk plotting support.

**Comment by fievelk** [comment]
If you want, you can test that `pip` currently installs NLTK without required dependencies:
```bash
$> virtualenv --python=/usr/bin/python3.4 test_env
$> source test_env/bin/activate
$> pip freeze
# Output is empty, since we did not install any module yet
$> pip install -U nltk
$> pip freeze
```
In fact, the version of `setup.py` which is being shipped with NLTK 3.2.1 is commented out: see [nltk/setup.py][1].

Instead, if we modify `install_requires` to be something like:
```python
install_requires = [
        'six>=1.10.0',
        'nose>=1.3.0',
        'tox>=1.6.1',
        'coverage>=3.7.1',
        'pylint>=1.1.0',
        'numpy>=1.8.0',
        'scipy>=0.13.2',
        'matplotlib>=1.3.1',
        'scikit-learn>=0.14.1',
        'python-crfsuite>=0.8.2',
        'gensim>=0.11.1',
        'pyparsing>=2.0.3',
        'twython>=3.2.0'
    ],
```
and we install NLTK using this updated version, we get all requirements:
```bash
$> pip install -e nltk/
$> pip freeze
astroid==1.4.8
boto==2.43.0
bz2file==0.98
coverage==4.2
cycler==0.10.0
gensim==0.13.3
isort==4.2.5
lazy-object-proxy==1.2.2
matplotlib==1.5.3
mccabe==0.5.2
nose==1.3.7
numpy==1.11.2
oauthlib==2.0.1
pluggy==0.4.0
py==1.4.31
pylint==1.6.4
pyparsing==2.1.10
python-crfsuite==0.8.4
python-dateutil==2.6.0
pytz==2016.7
requests==2.12.3
requests-oauthlib==0.7.0
scikit-learn==0.18.1
scipy==0.18.1
six==1.10.0
smart-open==1.3.5
tox==2.5.0
twython==3.4.0
virtualenv==15.1.0
wrapt==1.10.8
```
[1]: https://github.com/nltk/nltk/blob/991f2cd1e31f7c1ad144589ab4d2c76bee05aa7b/setup.py

**Comment by fievelk** [comment]
@kmike  I really like your solution.
In case we decide to follow this solution we should be careful not to break anything in the process. We should also add specific instructions to the existing NLTK documentation.

**Comment by kmike** [comment]
Adding all these packages to `install_requires` sounds like a maintenance nightmare - user must be able to build all these packages (many of them contain C/C++ extensions) in order to install NLTK - even if they are not needed at all for the user.

I don't see them as required dependencies, they are optional dependencies which enables some of NLTK features. Currently the only required dependency is six - a small pure Python module. There is no reason for users to build gensim if they only want to use nltk tokenizer.

Nose, tox, coverage and pylint (and a few other packages which are their transitive dependencies) are not NLTK dependencies, they are packages required to run NLTK tests. They are not needed for the end users.

**Comment by fievelk** [comment]
@kmike I was actually thinking about the whole thing from a developer's perspective and I was clearly overlooking more important aspects. Thank you for pointing that out :+1: 
In this case I suppose that dependencies do not need to be updated, so this issue is not an actual issue. @alvations do you agree on this? Do you have other ideas?

However, I like the idea of adding `extras_require` to `setup.py` in order to allow users to install optional dependencies.

**Comment by alvations** [comment]
I like @kmike suggestion too, having something like:

```
pip install nltk [plot] # Plotting + Tkinter + visual stuff included
```

and maybe:

```
pip install nltk [machine-learning] # All ML batteries included
```

and for lazy users:

```
pip install nltk [all] # All batteries included
```

----

On a related note, I think we need to also include the downloads in the `setup.py` code such that users won't see an error when using "NLTK default" functions e.g. :

```
from nltk import sent_tokenize, word_tokenize, ne_chunk, pos_tag
```

Currently, now the users must download:

```
import download
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
```

**Comment by stevenbird** [comment]
Thanks for all this – it will have a significant impact of people's experience getting set up with NLTK. A PR would be much appreciated. Also, if anyone needs the keys for pypi in order to update things there, please let me know.

**Comment by dimazest** [comment]
Including `extras_require` [1] into `setup.py` is be a huge step forward.

The `all` extra, that lists all the packages NLTK might ever require. Then the specific extras might be introduced.

[1] http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

**Comment by kmike** [comment]
`all` name has one caveat: it still won't get users non-Python dependencies like Stanford tools.

**Comment by stevenbird** [comment]
@alvations: is this resolved now?

**Comment by alvations** [comment]
Thanks @fievelk @dimazest @kmike , great fix on #1538!!

@stevenbird the main installation with `pip` is fixed.

Once the PyPI index is updated, I think `conda` will update their package on their manager too and that'll help quite a lot of users at installation.

----

One related issue is the automatic installation of the default (popular) models for `pos_tag` and `sent_tokenize` and `ne_chunk`. But maybe a separate issue to keep track of that will be more appropriate.

**Comment by alvations** [comment]
Resolved in #1538 and `popular` option for nltk/nltk_data#75
